src/EditorFeatures/Core/GoToImplementation/GoToImplementationCommandHandler.cs

```diff
@@ -81,28 +85,30 @@ public void ExecuteCommand(GoToImplementationCommandArgs args, Action nextHandle
             {
                 // We have all the cheap stuff, so let's do expensive stuff now
                 string messageToShow = null;
-                _waitIndicator.Wait(
-                    EditorFeaturesResources.Go_To_Implementation,
-                    EditorFeaturesResources.Locating_implementations,
-                    allowCancel: true,
-                    action: context =>
+
+                using (context.WaitContext.AddScope(allowCancellation: true, EditorFeaturesResources.Locating_implementations))
+                {
+                    var userCancellationToken = context.WaitContext.UserCancellationToken;
+                    if (canUseStreamingWindow)
                     {
-                        if (canUseStreamingWindow)
-                        {
-                            StreamingGoToImplementation(
-                                document, caretPosition,
-                                streamingService, streamingPresenter,
-                                context.CancellationToken, out messageToShow);
-                        }
-                        else
-                        {
-                            synchronousService.TryGoToImplementation(
-                                document, caretPosition, context.CancellationToken, out messageToShow);
-                        }
-                    });
+                        StreamingGoToImplementation(
+                            document, caretPosition,
+                            streamingService, streamingPresenter,
+                            userCancellationToken, out messageToShow);
+                    }
+                    else
+                    {
+                        synchronousService.TryGoToImplementation(
+                            document, caretPosition, userCancellationToken, out messageToShow);
+                    }
+                }
 
                 if (messageToShow != null)
                 {
+                    // We are about to show a modal UI dialog so we should take over the command execution
+                    // wait context. That means the command system won't attempt to show its own wait dialog 
+                    // and also will take it into consideration when measuring command handling duration.
+                    context.WaitContext.TakeOwnership();
                     var notificationService = document.Project.Solution.Workspace.Services.GetService<INotificationService>();
                     notificationService.SendNotification(messageToShow,
                         title: EditorFeaturesResources.Go_To_Implementation,
```

Comment:
it would be kind of nice if the notification system was somehow integrated into the wait context system.  But i imagine that could be really tough to do.

Reply:
I plan some additions to the editor's wait context system with a goal of deprecating Roslyn's IWaitContext stuff. That requires some new extensibility points to let Roslyn attach its IGlobalOperationNotificationService and this seems to be related task.
I'm tracking it via editor bug https://devdiv.visualstudio.com/DevDiv/_workitems/edit/550475